File: bilmarknad-mcp/src/bilmarknad_mcp/wayke.py
from __future__ import annotations

import os
from typing import Any

import httpx

from bilmarknad_mcp.schema import CarListing
from bilmarknad_mcp.soh import apply_soh

WAYKE_REST = "https://api.wayke.se/vehicles"
WAYKE_GQL = "https://gql.wayke.se/query"

GQL_SEARCH = """
query SearchVehicles($search: String, $skip: Int, $take: Int) {
  vehicles(search: $search, skip: $skip, take: $take) {
    id
    title
    shortDescription
    manufacturer { name }
    modelSeries { name }
    modelYear
    mileage
    price
    fuelType
    gearbox
    city
    organization { name }
    url
    image { url }
    registrationNumber
    published
  }
}
"""

GQL_VEHICLE = """
query GetVehicle($id: String!) {
  vehicle(id: $id) {
    id
    title
    shortDescription
    description
    manufacturer { name }
    modelSeries { name }
    modelYear
    mileage
    price
    fuelType
    gearbox
    city
    organization { name }
    url
    image { url }
    registrationNumber
    published
    data {
      properties
      options
    }
  }
}
"""


def _parse_vehicle(item: dict[str, Any]) -> CarListing:
    manufacturer = item.get("manufacturer") or {}
    series = item.get("modelSeries") or {}
    org = item.get("organization") or {}
    image = item.get("image") or {}
    mileage = item.get("mileage")
    price = item.get("price")
    listing = CarListing(
        source="wayke",
        id=str(item.get("id") or ""),
        title=item.get("title") or "",
        make=manufacturer.get("name"),
        model=series.get("name"),
        year=item.get("modelYear"),
        mileage_km=int(mileage) if mileage is not None else None,
        price_sek=int(price) if price is not None else None,
        fuel=item.get("fuelType"),
        transmission=item.get("gearbox"),
        location=item.get("city"),
        dealer_name=org.get("name"),
        url=item.get("url"),
        image_url=image.get("url"),
        published_at=item.get("published"),
        registration_number=item.get("registrationNumber"),
        raw=item,
    )
    apply_soh(listing, item.get("title"), item.get("shortDescription"), source="wayke_search")
    return listing

def _soh_detail_fields(item):
    fields = [item.get("description"), item.get("shortDescription")]
    data = item.get("data") or {}
    props = data.get("properties") or []
    opts = data.get("options") or []
    for coll in (props, opts):
        if isinstance(coll, list):
            for entry in coll:
                if isinstance(entry, dict):
                    fields.append(str(entry.get("name") or entry.get("label") or entry.get("value") or ""))
                else:
                    fields.append(str(entry))
        elif isinstance(coll, dict):
            for k, v in coll.items():
                fields.append(f"{k}: {v}")
    return [f for f in fields if f]


def _enrich_vehicle_soh(listing, item):
    apply_soh(listing, *_soh_detail_fields(item), source="wayke_detail")
    listing.raw = {**listing.raw, "detail": item}
    return listing


class WaykeClient:
    def __init__(self, client: httpx.Client | None = None, api_key: str | None = None):
        self._client = client
        self._owns = client is None
        self._api_key = api_key or os.environ.get("WAYKE_API_KEY")

    def _get_client(self) -> httpx.Client:
        if self._client is None:
            self._client = httpx.Client(timeout=30.0, headers={"User-Agent": "bilmarknad-mcp/0.1"})
        return self._client

    def search(
        self,
        q: str | None = None,
        rows: int = 20,
        page: int = 1,
    ) -> list[CarListing]:
        if self._api_key:
            rest = self._search_rest(q=q, rows=rows, page=page)
            if rest is not None:
                return rest
        return self._search_gql(q=q, rows=rows, page=page)

    def _search_rest(self, q, rows, page):
        client = self._get_client()
        headers = {"Authorization": f"Bearer {self._api_key}"}
        params: dict[str, Any] = {"take": rows, "skip": (page - 1) * rows}
        if q:
            params["search"] = q
        response = client.get(WAYKE_REST, params=params, headers=headers)
        if response.status_code == 401:
            return None
        response.raise_for_status()
        payload = response.json()
        items = payload if isinstance(payload, list) else payload.get("vehicles", payload.get("data", []))
        return [_parse_vehicle(item) for item in items]

    def _search_gql(self, q, rows, page):
        client = self._get_client()
        variables = {"search": q or "", "skip": (page - 1) * rows, "take": rows}
        response = client.post(
            WAYKE_GQL,
            json={"query": GQL_SEARCH, "variables": variables},
            headers={"Content-Type": "application/json"},
        )
        if response.status_code >= 400:
            return []
        data = response.json()
        vehicles = (((data.get("data") or {}).get("vehicles")) or [])
        return [_parse_vehicle(item) for item in vehicles]


    def get_vehicle(self, vehicle_id: str) -> CarListing | None:
        client = self._get_client()
        if self._api_key:
            response = client.get(f"{WAYKE_REST}/{vehicle_id}", headers={"Authorization": f"Bearer {self._api_key}"})
            if response.status_code == 200:
                item = response.json()
                listing = _parse_vehicle(item)
                return _enrich_vehicle_soh(listing, item)
        response = client.post(
            WAYKE_GQL,
            json={"query": GQL_VEHICLE, "variables": {"id": vehicle_id}},
            headers={"Content-Type": "application/json"},
        )
        if response.status_code >= 400:
            return None
        data = response.json()
        item = ((data.get("data") or {}).get("vehicle"))
        if not item:
            return None
        listing = _parse_vehicle(item)
        return _enrich_vehicle_soh(listing, item)

    def close(self):
        if self._owns and self._client is not None:
            self._client.close()

File: internal/bilmarknad/wayke/wayke.go
package wayke

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"

	"github.com/flyhard/swedish-car-mcp/internal/bilmarknad/httputil"
	"github.com/flyhard/swedish-car-mcp/internal/bilmarknad/schema"
	"github.com/flyhard/swedish-car-mcp/internal/bilmarknad/soh"
)

const restURL = "https://api.wayke.se/vehicles"

// Client queries Wayke via REST (with dealer API key) or public website scrape.
type Client struct {
	httpClient *http.Client
	owns       bool
	apiKey     string
}

func NewClient(c *http.Client, apiKey string) *Client {
	if c == nil {
		c = httputil.NewClient()
	}
	if apiKey == "" {
		apiKey = os.Getenv("WAYKE_API_KEY")
	}
	return &Client{httpClient: c, owns: c.Transport == nil, apiKey: apiKey}
}

func parseVehicle(item map[string]any) schema.CarListing {
	manufacturer, _ := item["manufacturer"].(map[string]any)
	series, _ := item["modelSeries"].(map[string]any)
	org, _ := item["organization"].(map[string]any)
	image, _ := item["image"].(map[string]any)
	var mileageKM, priceSEK, year *int
	if v := intFrom(item["mileage"]); v != nil {
		mileageKM = v
	}
	if v := intFrom(item["price"]); v != nil {
		priceSEK = v
	}
	if v := intFrom(item["modelYear"]); v != nil {
		year = v
	}
	listing := schema.CarListing{
		Source:             "wayke",
		ID:                 fmt.Sprint(item["id"]),
		Title:              str(item["title"]),
		Make:               strPtr(manufacturer, "name"),
		Model:              strPtr(series, "name"),
		Year:               year,
		MileageKM:          mileageKM,
		PriceSEK:           priceSEK,
		Fuel:               strPtr(item, "fuelType"),
		Transmission:       strPtr(item, "gearbox"),
		Location:           strPtr(item, "city"),
		DealerName:         strPtr(org, "name"),
		URL:                strPtr(item, "url"),
		ImageURL:           strPtr(image, "url"),
		PublishedAt:        strPtr(item, "published"),
		RegistrationNumber: strPtr(item, "registrationNumber"),
		Raw:                item,
	}
	soh.Apply(&listing, "wayke_search", listing.Title, str(item["shortDescription"]))
	return listing
}

func detailFields(item map[string]any) []string {
	fields := []string{str(item["description"]), str(item["shortDescription"])}
	data, _ := item["data"].(map[string]any)
	for _, key := range []string{"properties", "options"} {
		coll := data[key]
		switch typed := coll.(type) {
		case []any:
			for _, entry := range typed {
				if m, ok := entry.(map[string]any); ok {
					fields = append(fields, str(m["name"]), str(m["label"]), str(m["value"]))
				} else {
					fields = append(fields, fmt.Sprint(entry))
				}
			}
		case map[string]any:
			for k, v := range typed {
				fields = append(fields, fmt.Sprintf("%s: %v", k, v))
			}
		}
	}
	out := make([]string, 0, len(fields))
	for _, f := range fields {
		if strings.TrimSpace(f) != "" {
			out = append(out, f)
		}
	}
	return out
}

func enrichSOH(listing *schema.CarListing, item map[string]any) *schema.CarListing {
	soh.Apply(listing, "wayke_detail", detailFields(item)...)
	if listing.Raw == nil {
		listing.Raw = map[string]any{}
	}
	listing.Raw["detail"] = item
	return listing
}

func (c *Client) Search(ctx context.Context, q *string, rows, page int) ([]schema.CarListing, error) {
	if c.apiKey != "" {
		rest, err := c.searchREST(ctx, q, rows, page)
		if err == nil && rest != nil {
			return rest, nil
		}
	}
	return c.searchScrape(ctx, q, rows, page)
}

func (c *Client) searchREST(ctx context.Context, q *string, rows, page int) ([]schema.CarListing, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, restURL, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+c.apiKey)
	query := req.URL.Query()
	query.Set("take", fmt.Sprint(rows))
	query.Set("skip", fmt.Sprint((page-1)*rows))
	if q != nil && *q != "" {
		query.Set("search", *q)
	}
	req.URL.RawQuery = query.Encode()
	resp, err := httputil.DoWithRetry(ctx, c.httpClient, req, "wayke", httputil.DefaultRetryPolicy())
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode == 401 {
		return nil, nil
	}
	if resp.StatusCode >= 400 {
		return nil, fmt.Errorf("wayke rest: %s", resp.Status)
	}
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	var payload any
	if err := json.Unmarshal(body, &payload); err != nil {
		return nil, err
	}
	var items []any
	switch typed := payload.(type) {
	case []any:
		items = typed
	case map[string]any:
		if v, ok := typed["vehicles"].([]any); ok {
			items = v
		} else if v, ok := typed["data"].([]any); ok {
			items = v
		}
	}
	out := make([]schema.CarListing, 0, len(items))
	for _, item := range items {
		if m, ok := item.(map[string]any); ok {
			out = append(out, parseVehicle(m))
		}
	}
	return out, nil
}

func (c *Client) GetVehicle(ctx context.Context, vehicleID string) (*schema.CarListing, error) {
	if c.apiKey != "" {
		req, err := http.NewRequestWithContext(ctx, http.MethodGet, restURL+"/"+vehicleID, nil)
		if err == nil {
			req.Header.Set("Authorization", "Bearer "+c.apiKey)
			resp, err := httputil.DoWithRetry(ctx, c.httpClient, req, "wayke", httputil.DefaultRetryPolicy())
			if err == nil {
				defer resp.Body.Close()
				if resp.StatusCode == 200 {
					body, _ := io.ReadAll(resp.Body)
					var item map[string]any
					if json.Unmarshal(body, &item) == nil {
						listing := parseVehicle(item)
						return enrichSOH(&listing, item), nil
					}
				}
			}
		}
	}
	return c.getVehicleScrape(ctx, vehicleID)
}

func (c *Client) Close() {
	if c.owns && c.httpClient != nil {
		c.httpClient.CloseIdleConnections()
	}
}

func str(v any) string {
	if v == nil {
		return ""
	}
	return fmt.Sprint(v)
}

func strPtr(m map[string]any, key string) *string {
	if m == nil {
		return nil
	}
	v, ok := m[key]
	if !ok || v == nil {
		return nil
	}
	s := strings.TrimSpace(fmt.Sprint(v))
	if s == "" {
		return nil
	}
	return &s
}

func intFrom(v any) *int {
	if v == nil {
		return nil
	}
	switch n := v.(type) {
	case float64:
		i := int(n)
		return &i
	case int:
		return &n
	default:
		var i int
		if _, err := fmt.Sscan(fmt.Sprint(v), &i); err != nil {
			return nil
		}
		return &i
	}
}
